python/ray/workflow/api.py

```diff
@@ -168,37 +194,36 @@ def run_async(
         # Workflow ID format: {Entry workflow UUID}.{Unix time to nanoseconds}
         workflow_id = f"{str(uuid.uuid4())}.{time.time():.9f}"
 
+    workflow_manager = workflow_access.get_management_actor()
+    if ray.get(workflow_manager.is_workflow_non_terminating.remote(workflow_id)):
+        raise RuntimeError(f"Workflow '{workflow_id}' is already running or pending.")
+
     state = workflow_state_from_dag(dag, input_data, workflow_id)
     logger.info(f'Workflow job created. [id="{workflow_id}"].')
-
     context = workflow_context.WorkflowStepContext(workflow_id=workflow_id)
     with workflow_context.workflow_step_context(context):
         # checkpoint the workflow
-        ws = WorkflowStorage(workflow_id)
-        ws.save_workflow_user_metadata(metadata)
-
-        job_id = ray.get_runtime_context().job_id.hex()
-
-        try:
-            ws.get_entrypoint_step_id()
-            wf_exists = True
-        except Exception:
-            # The workflow does not exist. We must checkpoint entry workflow.
-            ws.save_workflow_execution_state("", state)
-            wf_exists = False
-        workflow_manager = workflow_access.get_management_actor()
-        if ray.get(workflow_manager.is_workflow_non_terminating.remote(workflow_id)):
-            raise RuntimeError(
-                f"Workflow '{workflow_id}' is already running or pending."
-            )
+        @client_mode_wrap
+        def _try_checkpoint_workflow(workflow_state) -> bool:
+            ws = WorkflowStorage(workflow_id)
+            ws.save_workflow_user_metadata(metadata)
+            try:
+                ws.get_entrypoint_step_id()
+                return True
+            except Exception:
+                # The workflow does not exist. We must checkpoint entry workflow.
+                ws.save_workflow_execution_state("", workflow_state)
+                return False
+
+        wf_exists = _try_checkpoint_workflow(state)
         if wf_exists:
             return resume_async(workflow_id)
-        ignore_existing = ws.load_workflow_status() == WorkflowStatus.NONE
         ray.get(
             workflow_manager.submit_workflow.remote(
-                workflow_id, state, ignore_existing=ignore_existing
+                workflow_id, state, ignore_existing=False
```

Comment:
any reason we put ignore_existing=False?

Reply:
The original logic is equivalent to `ignore_existing=False`, this double checks we would not run an existing workflow again. `ignore_existing=True` is being used inside the workflow management actor when caling `resume_async(workflow_id)`